### PR TITLE
Un-delay opening package objects when early-completing a (different) package object symbol

### LINKED ISSUE #12740: not found: type Serializable after Scala 2.13.6 upgrade up to 2.13.7+
## Reproduction steps

**I was unable to get a locally reproducible / relatively easy project that fails.** ~I could not reproduce it locally at all. 😭~ Thx team with all the help to reproduce it!

Scala version: **2.13.7+**
**SBT and JDK versions independent.**

Code:
```scala
package test

package object X extends Serializable { }

object Test {
  def main(args: Array[String]): Unit = {
    
  }

}
```

Scasite: https://scastie.scala-lang.org/pomadchin/X8YFi5YuT9yUNTp1fpmNEA/1

Publish: `$ ./sbt +publishLocal`

Steps to _try_ to reproduce locally with the https://github.com/PDAL/java project

```scala
$ git clone git@github.com:PDAL/java.git pdal-java
$ git checkout -f bug/scala-upd
$ cd pdal-java
$ docker run --rm -it -v $PWD:/pdal daunnc/pdal-ubuntu:2.5.1 bash
$ cd /pdal
$ ./sbt ++2.13 core-scala/publishLocal
```

I know that it makes this bug / feature report very hard to work with, but mb there is any chance to get some help tracing it?
Any help / comments are really appreciated.

I spent a the whole weekend trying to catch it and was only able ("got lucky") to catch it in a smaller OSS project and reduce failure await times in CI.

## Problem

In several projects (https://github.com/PDAL/java, https://github.com/locationtech/geotrellis/) after updating Scala 2.13.6 up to 2.13.7+ I get `not found: type Serializable` error on the package `+publish` command invocation, more specifically, on the `Compile/doc` stage. 

The most unclear thing is that it may fail within some specific modules of a project, not in every module of a project, I was unable to catch the real issue / to find a simple reproducible way to catch it.

By experiments, I was able to double check that this behavior is not related to the SBT version and to the JDK version.

**Code:** https://github.com/PDAL/java/blob/bug/scala-upd/core-scala/src/main/scala/io/pdal/pipeline/package.scala#L19
**Diff that triggers the CI error:** https://github.com/PDAL/java/commit/f068ac26d59f893dce899dbe77e192f09d82400d
**CI Error:** https://github.com/PDAL/java/actions/runs/4277401013/jobs/7446185708#step:6:201

```scala
[error] /__w/java/java/core-scala/src/main/scala/io/pdal/pipeline/package.scala:19:75: not found: type Serializable
[error] package object pipeline extends json.Implicits with syntax.Implicits with Serializable
[error]                                                                           ^
[error] one error found
...
[error] Scaladoc generation failed
[error] 	at xsbt.Runner.run(ScaladocBridge.scala:50)
[error] 	at xsbt.ScaladocBridge.run(ScaladocBridge.scala:21)
[error] 	at sbt.internal.inc.AnalyzingCompiler.doc(AnalyzingCompiler.scala:154)
[error] 	at sbt.internal.inc.AnalyzingCompiler.doc(AnalyzingCompiler.scala:133)
[error] 	at sbt.Doc$.$anonfun$scaladoc$1(Doc.scala:52)
[error] 	at sbt.Doc$.$anonfun$scaladoc$1$adapted(Doc.scala:40)
[error] 	at sbt.RawCompileLike$.$anonfun$prepare$1(RawCompileLike.scala:79)
[error] 	at sbt.RawCompileLike$.$anonfun$prepare$1$adapted(RawCompileLike.scala:72)
[error] 	at sbt.RawCompileLike$.$anonfun$cached$4(RawCompileLike.scala:63)
[error] 	at sbt.RawCompileLike$.$anonfun$cached$4$adapted(RawCompileLike.scala:61)
[error] 	at sbt.util.Tracked$.$anonfun$inputChangedW$1(Tracked.scala:219)
[error] 	at sbt.RawCompileLike$.$anonfun$cached$1(RawCompileLike.scala:68)
[error] 	at sbt.RawCompileLike$.$anonfun$cached$1$adapted(RawCompileLike.scala:52)
[error] 	at sbt.Defaults$.$anonfun$docTaskSettings$4(Defaults.scala:2156)
[error] 	at scala.Function1.$anonfun$compose$1(Function1.scala:49)
[error] 	at sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:62)
[error] 	at sbt.std.Transform$$anon$4.work(Transform.scala:68)
[error] 	at sbt.Execute.$anonfun$submit$2(Execute.scala:282)
[error] 	at sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:23)
[error] 	at sbt.Execute.work(Execute.scala:[291])
[error] 	at sbt.Execute.$anonfun$submit$1(Execute.scala:282)
[error] 	at sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:265)
[error] 	at sbt.CompletionService$$anon$2.call(CompletionService.scala:64)
[error] 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
[error] 	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
[error] 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
[error] 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
[error] 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
[error] 	at java.base/java.lang.Thread.run(Thread.java:829)
[error] (core-scala / Compile / doc) Scaladoc generation failed
```

Workarounds that work for now: 
* Skipping the doc stage 
* Using java.io.Serializable instead of scala.Serializable in places that cause problems 
* Use Scala 2.13.6

Ref to another project issue where I have the same problem https://github.com/locationtech/geotrellis/issues/3503

### PR DESCRIPTION
**Pull request by som-snytt**
Fixes scala/bug#12740

### CODE REVIEW

**Comment by pomadchin** [comment]
Checked this PR on my failing projects, it's all 🔥

**Comment by som-snytt** [comment]
Does that mean I burned it all to the ground?

I will take a second look at whether this is the right fix, I currently have restricted capacity. Not sure why roots are different or what the lazy packages PR was fixing.

**Comment by pomadchin** [comment]
> Does that mean I burned it all to the ground?

Haha, sorry; it looks like with this PR all previously failing projects are green now. I can check one more time later in a more complex project as well.

**Comment by SethTisue** [comment]
let's see what the community build thinks of 2.13.11-bin-65041ac-SNAPSHOT :  ~https://scala-ci.typesafe.com/job/scala-2.13.x-jdk11-integrate-community-build/4323/~ https://scala-ci.typesafe.com/job/scala-2.13.x-jdk11-integrate-community-build/4327/ queued

**Comment by lrytz** [comment]
There's always a risk of causing cycles when changing the compiler to force more things early...

> Not sure why roots are different

Other imports of packages cause initialization of the package symbol?

> what the lazy packages PR was fixing

So this PR reverts some earlier change?

**Comment by SethTisue** [comment]
scala-java8-compat is failing at https://scala-ci.typesafe.com/job/scala-2.13.x-jdk11-integrate-community-build/4329/ , but perhaps it isn't this PR's fault — let's see if the same failure also shows up at https://github.com/scala/community-build/pull/1646

**Comment by som-snytt** [comment]
@SethTisue I see the error also after commenting out this change.
```
[error] Test scala.compat.java8.StepperTest.spliterating failed: java.lang.NoSuchMethodError: 'java.util.Spliterator$OfInt scala.collection.IntStepper.spliterator$mcI$sp$(scala.collection.IntStepper)', took 0.002 sec
```

**Comment by SethTisue** [comment]
~Indeed, the other community build run shows that some other PR is the culprit. I'll pursue that separately.~

**Comment by SethTisue** [comment]
> Indeed, the other community build run shows that some other PR is the culprit

gah, I must have gotten confused, with all the runs I was doing — the scala-java8-compat failure did _not_ occur on https://github.com/scala/community-build/pull/1646

I don't know how to reconcile that with

> I see the error also after commenting out this change

try rebasing on top of 0fd61f7, perhaps?

**Comment by SethTisue** [comment]
Oh, the scala-java8-compat test in question (`Issue247Test`) is an extremely weird test — it uses coursier to retrieve artifacts at runtime. I'll just disable it in the community build.

**Comment by som-snytt** [comment]
I just ran `test` on my JVM 19, what could go wrong.

**Comment by SethTisue** [comment]
@som-snytt if you rebase, I'll run the community build again

**Comment by som-snytt** [comment]
I'll rework this. When `packageobjects` types the parent, `scala.package` is not opened to satisfy look-up. Type aliases for classes in `java.lang` resolve there, but `Serializable` in `java.io` and `NoSuchElementException` in `java.util` are not found. It also seems to me that the same mechanism should be used for forcing (and erroring) user roots as well as "default" roots.

File: src/compiler/scala/tools/nsc/typechecker/Contexts.scala [hunk 1]
```diff
@@ -1898,6 +1898,7 @@ trait Contexts { self: Analyzer =>
 
     /** If requireExplicit is true, wildcard imports are not considered. */
     def importedSelectedSymbol(name: Name, requireExplicit: Boolean): (ImportSelector, Symbol) = {
+      if (isRootImport && qual.symbol.hasPackageFlag) openPackageModule(qual.symbol, force = true)
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/typechecker/Contexts.scala]: Why do we need a special case for root imports, do you know?

I tried two files
```
package a {
  package object p {
    type Sr = java.io.Serializable
  }
}
```

```
import a.p._
package b {
  package object q extends Sr
}
```

When compiling the second, `openPackageModule` for `a.b` is called in `PackageLoader.doComplete`. Ideas why this doesn't happen for root imports?

Comment (som-snytt) [reply src/compiler/scala/tools/nsc/typechecker/Contexts.scala]: Root imports don't benefit from namer. The nicer place to set this up is when creating the root context: intuitively, do the work when completing the lazy type. I don't see yet how to do that for `Predef`, but I'll continue fiddling.

Comment (lrytz) [reply src/compiler/scala/tools/nsc/typechecker/Contexts.scala]: So the `PackageLoader` for `scala` is forced very early (before namer) due to things in Definitions (`ObjectClass.initialize` triggers a bunch of stuff through the classfile parser).

This means that opening the package object is deferred for `scala`.

The lookup for `Serializable` when compiling `package object XX extends Serializable` happens in phase `packageobjects`, which does

```
        openPackageObjectsTraverser(unit.body)     // triggers lookup
        deferredOpen.foreach(openPackageModule(_)) // too late
```

This order is deliberate though (https://github.com/scala/scala/pull/9661/files).

The "root imports" focus of this PR is probably a red herring? I can compile `test/files/pos/t12740b` below using 2.13.12 with `-Yimports` without problem.

Mabye we just need to force the `scala` package early? Here's the list of packages that are completed on initialization (even before parser):
```
<root>
java
java.lang
jdk
jdk.internal
jdk.internal.vm
jdk.internal.vm.annotation
> scala
scala.annotation
scala.runtime
> scala.collection
> scala.collection.immutable
java.io
java.lang.constant
```

Then two more during namer

```
> scala.reflect
scala.annotation.meta
```

I marked the ones that have a non-empty package object.

**Comment by lrytz** [comment]
@som-snytt rebased (without changing your commits) and pushed an attempt according to https://github.com/scala/scala/pull/10333#discussion_r1402293111

File: src/compiler/scala/tools/nsc/typechecker/Contexts.scala (hunk 1 above)
Comment (lrytz) [reply src/compiler/scala/tools/nsc/typechecker/Contexts.scala]: So I think my attempt could partially subvert the fix added in https://github.com/scala/scala/pull/9661. According to the PR description, if we had

`package object scala extends SP`

and we'd re-compile the source file where `SP` is defined but not the pacakge object, we could run into a spurious "symbol not found" error.

But the few package objects mentioned above are are all without parent, and anyway they are all in our source code, so my commit is not reverting #9661 for user code.

Comment (som-snytt) [reply src/compiler/scala/tools/nsc/typechecker/Contexts.scala]: I haven't had a chance to revisit this and cope with the "red herring" remark. To me, this was about a uniform mechanism for `-Yimports`.

**Comment by som-snytt** [comment]
@lrytz thanks for your fix and the improved title. I will also experiment whether user-root-import-forcing can now just use this mechanism. I slept last night instead of pursuing it.

Edit: reminder to self that it could error earlier on qual.tpe.isError

**Pull request merged**